libs/vm/src/analyser.cpp

```diff
@@ -357,25 +390,438 @@ bool Analyser::Analyse(BlockNodePtr const &root, std::vector<std::string> &error
+{
+  auto const &function_name_node = ConvertToExpressionNodePtr(function_node->children[1]);
+  auto const &annotations        = function_node->children[0];
+
+  bool is_action = false;
+  bool is_init   = false;
+  bool is_query  = false;
+
+  assert(!annotations || annotations->node_kind == NodeKind::Annotations);
+
+  switch (function_node->node_kind)
+  {
+  case NodeKind::ContractFunction:
+  {
+    if (HasAnnotation(annotations, ACTION_ANNOTATION))
+    {
+      is_action = true;
+    }
+    else
+    {
+      AddError(function_name_node->line, "Contract functions must carry the @action annotation");
+    }
+  }
+  break;
+  case NodeKind::FreeFunctionDefinition:
+  {
+    if (HasAnnotation(annotations, ACTION_ANNOTATION))
+    {
+      is_action = true;
+    }
+    if (HasAnnotation(annotations, INIT_ANNOTATION))
+    {
+      is_init = true;
+      metadata.init_functions.emplace_back(function_node, filename_);
+    }
+    if (HasAnnotation(annotations, QUERY_ANNOTATION))
+    {
+      is_query = true;
+    }
+    if (HasAnnotation(annotations, CLEAR_ANNOTATION))
+    {
+      metadata.clear_functions.emplace_back(function_node, filename_);
+    }
+    if (HasAnnotation(annotations, OBJECTIVE_ANNOTATION))
+    {
+      metadata.objective_functions.emplace_back(function_node, filename_);
+    }
+    if (HasAnnotation(annotations, PROBLEM_ANNOTATION))
+    {
+      metadata.problem_functions.emplace_back(function_node, filename_);
+    }
+    if (HasAnnotation(annotations, WORK_ANNOTATION))
+    {
+      metadata.work_functions.emplace_back(function_node, filename_);
+    }
+  }
+  break;
+  case NodeKind::MemberFunctionDefinition:
+  {
+    if (annotations)
+    {
+      AddError(annotations->line, "Annotations on member functions are not permitted");
+    }
+  }
+  break;
+  default:
+    assert(false);
+    break;
+  }
+
+  auto const &return_type = function_name_node->function->return_type != nullptr
+                                ? function_name_node->function->return_type
+                                : void_type_;
+
+  if (is_action)
+  {
+    if (return_type != int64_type_ && return_type != void_type_)
+    {
+      AddError(function_name_node->line, "@action functions must either be void or return Int64");
+    }
+  }
+
+  if (is_init)
+  {
+    if (return_type != int64_type_ && return_type != void_type_)
+    {
+      AddError(function_name_node->line, "@init functions must either be void or return Int64");
+    }
+
+    auto const param_types = function_name_node->function->parameter_types;
+    if (!param_types.empty() && !(param_types.size() == 1 && param_types.back() == address_type_))
+    {
+      AddError(function_name_node->line,
+               "@init functions must accept eithoer no parameters or exactly one parameter of type "
+               "Address");
+    }
+  }
+
+  if (is_query)
+  {
+    if (return_type == void_type_)
+    {
+      AddError(function_name_node->line, "@query functions must not be void");
```

Comment:
"must return a value" ?

Reply:
I've gone with "must have a return type". If they forget to return when the function's not void, they'll get a meaningful error on the next attempt